core/src/main/java/org/pentaho/di/core/encryption/CertificateGenEncryptUtil.java

```diff
@@ -0,0 +1,136 @@
+      "Certificate Encryption Utility", LoggingObjectType.GENERAL, null );
+  private static final LogChannel log = new LogChannel( loggingObject );
+
+  public static KeyPair generateKeyPair() {
+    KeyPair pair = null;
+    try {
+      KeyPairGenerator keyPairGen = KeyPairGenerator.getInstance( PUBLIC_KEY_ALGORITHM );
+      keyPairGen.initialize( KEY_SIZE );
+      pair = keyPairGen.generateKeyPair();
+    } catch ( Exception ex ) {
+      log.logError( ex.getLocalizedMessage(), ex );
+    }
+    return pair;
+  }
+
+  public static Key generateSingleKey() throws NoSuchAlgorithmException {
+    Key key = KeyGenerator.getInstance( SINGLE_KEY_ALGORITHM ).generateKey();
+    return key;
+  }
+
+  public static byte[] encodeKeyForTransmission( Key encodingKey, Key keyToEncode ) throws NoSuchAlgorithmException,
+    NoSuchPaddingException, InvalidKeyException, IllegalBlockSizeException {
+    Cipher cipher = Cipher.getInstance( TRANSMISSION_CIPHER_PARAMS );
+    cipher.init( Cipher.WRAP_MODE, encodingKey );
+    byte[] encodedKey = cipher.wrap( keyToEncode );
+    return encodedKey;
+  }
+
+  public static Key decodeTransmittedKey( byte[] sessionKey, byte[] transmittedKey, boolean privateKey )
+    throws InvalidKeySpecException, NoSuchAlgorithmException, NoSuchPaddingException, InvalidKeyException {
+    KeySpec keySpec = null;
+    Key keyKey = null;
+    if ( transmittedKey == null || sessionKey == null ) {
+      return null;
+    }
+    if ( !privateKey ) {
+      keySpec = new X509EncodedKeySpec( sessionKey );
+      keyKey = KeyFactory.getInstance( PUBLIC_KEY_ALGORITHM ).generatePublic( keySpec );
+    } else {
+      keySpec = new PKCS8EncodedKeySpec( sessionKey );
+      keyKey = KeyFactory.getInstance( PUBLIC_KEY_ALGORITHM ).generatePrivate( keySpec );
+    }
+    Cipher keyCipher = Cipher.getInstance( TRANSMISSION_CIPHER_PARAMS );
+    keyCipher.init( Cipher.UNWRAP_MODE, keyKey );
+    return keyCipher.unwrap( transmittedKey, SINGLE_KEY_ALGORITHM, Cipher.SECRET_KEY );
+  }
+
+  public static Cipher initDecryptionCipher( Key unwrappedKey, byte[] unencryptedKey ) throws NoSuchAlgorithmException,
+    NoSuchPaddingException, InvalidKeyException {
+    Cipher decryptionCip = Cipher.getInstance( SINGLE_KEY_ALGORITHM );
+    if ( unwrappedKey != null ) {
+      decryptionCip.init( Cipher.ENCRYPT_MODE, unwrappedKey );
+    } else {
+      SecretKeySpec sks = new SecretKeySpec( unencryptedKey, SINGLE_KEY_ALGORITHM );
+      decryptionCip.init( Cipher.ENCRYPT_MODE, sks );
+    }
+    return decryptionCip;
+  }
+
+  public static byte[] encryptUsingKey( byte[] data, Key key ) {
+    byte[] result = null;
+    try {
+      Cipher cipher = Cipher.getInstance( PUBLIC_KEY_ALGORITHM );
+      cipher.init( Cipher.ENCRYPT_MODE, key );
+      result = cipher.doFinal( data );
+    } catch ( Exception ex ) {
+      log.logError( ex.getLocalizedMessage(), ex );
+    }
+    return result;

```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 114:122, 126:134.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.